content/post/2020-11-18-testthat-utility-belt/index.Rmd

```diff
@@ -0,0 +1,106 @@
+- testing
+output: hugodown::hugo_document
+---
+
+```{r setup, include=FALSE}
+
+knitr::opts_chunk$set(fig.path = "", comment = "")
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+If your package uses testthat for unit testing, as [many packages on CRAN](https://www.tidyverse.org/blog/2019/11/testthat-2-3-0/), you might have wondered at least once about where to put "things" you wished to use in several tests: a function generating a random string, an example image, etc.
+In this blog post, we shall offer a round-up of how to use such code and files when testing your package.
+
+## Code called in your tests
+
+Remember our post about [internal functions in R packages](/2019/12/12/internal-functions/)?
+What about internal functions in _unit tests_ of R packages?
+And code that needs to be run for tests?
+
+Where to put your code and function depends on where you'll want to use them.
+
+* It's best not to touch `tests/testthat/test.R`.
+* R scripts under `tests/testthat/` whose name start with `setup-` are loaded before tests are run but not with `devtools::load_all()` which means the code in there is _not_ available when you're interactively debugging a test. And yes, you'll be interactively debugging tests more often than you wish. :wink:
+* R scripts under `tests/testthat/` whose name start with `helper-` are loaded with `devtools::load_all()` so they are available for both tests and interactive debugging. Now, it also means their behavior does not different from R scripts under `R/` so you might put them in that directory too as recommended in testthat docs. However, it also means they are installed with the package which might (slightly!) increase its size, and that they are with the rest of your package code which might [put you off](https://community.rstudio.com/t/why-are-tests-testthat-helper-files-discouraged-in-testthat/85253).
+
+To summarize,
+
+| File                           | Run before tests | Loaded via `load_all()` | Installed with the package[^installed] | Testable[^testable] |
+|--------------------------------|------------------|-------------------------|----------------------------|----------|
+| tests/testthat/setup-.\*.R     | ✔️               | \-                      | \-                         | \-       |
+| tests/testthat/helper-.\*.R    | ✔️               | ✔️                      | \-                         | \-       |
+| R/any-name.R                   | ✔️               | ✔️                      | ✔️                         | ✔️       |
+| tests/testthat/anything-else.R | \-               | \-                      | \-                         | \-       |
```

Comment:
I don't think you need the dashes for setup and helper files, so it is just `setup*.R`. Btw. is this a regexp? Then you need to escape the last dot. But it think it is easier to use globs for filenames, like the shells, i.e. `setup*.R`.